### ISSUE TITLE
odd result from uftrace replay --kernel on ARM32

### ISSUE THREAD

**Issue by iamwonseok** [namhyung/uftrace#78]
Hi, @namhyung 

I think that the result of uftrace with any example on arm32 is very similar to a result with the same  example on x86_64, basically. 

Log from `uftrace ./hello` on x86_64:
```
$ uftrace ./hello
Hello world
# DURATION    TID     FUNCTION
   1.124 us [23238] | __monstartup();
   0.667 us [23238] | __cxa_atexit();
            [23238] | main() {
   7.538 us [23238] |   printf();
   8.301 us [23238] | } /* main */
```

Log from `uftrace ./hello` on arm32:
```
$ uftrace ./hello -kernel
Hello world
# DURATION    TID     FUNCTION
  17.700 us [ 2337] | __cxa_atexit();
            [ 2337] | main() {
  73.500 us [ 2337] |   printf();
  89.600 us [ 2337] | } /* main */
```

The results are very similar to each other. By the way, using `--kernel` option, they show quite different results as below.

Log from `uftrace --kernel ./hello` on x86_64:
```
$ uftrace --kernel ./hello
Hello world
# DURATION    TID     FUNCTION
   1.040 us [23246] | __monstartup();
   0.650 us [23246] | __cxa_atexit();
            [23246] | main() {
            [23246] |   printf() {
            [23246] |     sys_newfstat() {
   0.974 us [23246] |       smp_irq_work_interrupt();
   4.759 us [23246] |     } /* sys_newfstat */
   3.685 us [23246] |     sys_mmap();
   3.773 us [23246] |     __do_page_fault();
   1.861 us [23246] |     smp_irq_work_interrupt();
  22.089 us [23246] |   } /* printf */
  22.880 us [23246] | } /* main */
```

Log from `uftrace --kernel ./hello` on arm32:
```
$ uftrace --kernel ./hello        
Hello world
# DURATION    TID     FUNCTION
   4.900 us [ 2342] | __cxa_atexit();
            [ 2342] | sys_clock_gettime() {
  20.800 us [ 2342] |   gic_handle_irq();
  45.600 us [ 2342] | } /* __cxa_atexit */
   5.000 us [ 2342] | main();
            [ 2342] | sys_clock_gettime() {
            [ 2342] |   printf() {
   4.900 us [ 2342] |   } /* sys_clock_gettime */
  15.600 us [ 2342] |   sys_fstat64();
  42.900 us [ 2342] |   sys_mmap_pgoff();
  48.500 us [ 2342] |   do_page_fault();
            [ 2342] |   sys_clock_gettime() {
   7.700 us [ 2342] |   } /* printf */
 176.200 us [ 2342] | } /* sys_clock_gettime */
  20.700 us [ 2342] | gic_handle_irq();
   8.200 us [ 2342] | sys_clock_gettime();
```

I think there is something wrong in arm32, but I don't know whether it is a bug or not. Could you please let me know how do I fix it? And I left some related-questions.

1. Why they are different entry point? Both of them have a similar `hello.sym` file.
2. I think `uftrace report` on arm shows odd result, too. IMHO, It seems that sys_clock_gettime() returns wrong result in record phase. So I was thinking the problems comes from arch_timer in armv7. By the way, `date` command returns correct result on my machine. 

`uftrace report` on arm:  It shows that __cxa_atexit spent most of the time in application.
```
$uftrace report
  Total time   Self time       Calls  Function
  ==========  ==========  ==========  ====================================
   37.600 us   19.700 us           1  __cxa_atexit
    8.000 us    8.000 us           2  sys_clock_gettime
    6.500 us    6.500 us           1  printf
    6.300 us    6.300 us           1  main
```

**Comment by namhyung** [comment]
What's your kernel version?  It depends on the implementation of the clock_gettime() syscall which uftrace uses internally.  On x86_64, it's implemented in VDSO so it doesn't go into the kernel, but VDSO support on ARM was added recently so your system might not support it yet.  If you want a similar result, I suggest you filter out the sys_clock_gettime kernel function.  Maybe it'd be better to detect it in uftrace and add the filter automatically..

```
$ uftrace -k -N sys_clock_gettime@kernel hello
```

**Comment by namhyung** [comment]
Yes, but changing kernel is hard so I prefer uftrace filtering it with
"VDSO is not found, filtering sys_clock_gettime" message if possible.

2017년 3월 24일 (금) 오후 5:36, iamwonseok <notifications@github.com>님이 작성:

**Comment by namhyung** [comment]
@iamwonseok I think you can check `/proc/self/maps` contains '[vdso]', and, if not, append `";!sys_clock_gettime@kernel` to opts->filter before calling `setup_kernel_tracing()` in the `command_record()`.

**Comment by iamwonseok** [comment]
@namhyung I ran the uftrace on my machin with two different kernel configurations(`CONFIG_VDSO=y` and `# CONFIG_VDSO is not set`). By the way, I got the same uftrace result in both cases.

case 1: VDSO is enabled:
```
root@kesl:~/arm/test# cat /proc/self/maps 
00008000-0000f000 r-xp 00000000 b3:19 160669     /bin/cat
00016000-00017000 r--p 00006000 b3:19 160669     /bin/cat
00017000-00018000 rw-p 00007000 b3:19 160669     /bin/cat
00018000-00039000 rw-p 00000000 00:00 0          [heap]
b6edc000-b6fb6000 r-xp 00000000 b3:19 176829     /lib/arm-linux-gnueabihf/libc-2.19.so
b6fb6000-b6fbd000 ---p 000da000 b3:19 176829     /lib/arm-linux-gnueabihf/libc-2.19.so
b6fbd000-b6fbf000 r--p 000d9000 b3:19 176829     /lib/arm-linux-gnueabihf/libc-2.19.so
b6fbf000-b6fc0000 rw-p 000db000 b3:19 176829     /lib/arm-linux-gnueabihf/libc-2.19.so
b6fc0000-b6fc3000 rw-p 00000000 00:00 0 
b6fc3000-b6fda000 r-xp 00000000 b3:19 176819     /lib/arm-linux-gnueabihf/ld-2.19.so
b6fdb000-b6fdc000 rw-p 00000000 00:00 0 
b6fe0000-b6fe2000 rw-p 00000000 00:00 0 
b6fe2000-b6fe3000 r--p 00017000 b3:19 176819     /lib/arm-linux-gnueabihf/ld-2.19.so
b6fe3000-b6fe4000 rw-p 00018000 b3:19 176819     /lib/arm-linux-gnueabihf/ld-2.19.so
bef42000-bef63000 rw-p 00000000 00:00 0          [stack]
bef74000-bef75000 r-xp 00000000 00:00 0          [sigpage]
bef75000-bef76000 r--p 00000000 00:00 0          [vvar]
bef76000-bef77000 r-xp 00000000 00:00 0          [vdso]
ffff0000-ffff1000 r-xp 00000000 00:00 0          [vectors]

root@kesl:~/arm/test# uftrace --kernel ./hello
Hello world
# DURATION    TID     FUNCTION
   2.600 us [ 2071] | __cxa_atexit();
   3.800 us [ 2071] | sys_clock_gettime();
   2.700 us [ 2071] | main();
            [ 2071] | sys_clock_gettime() {
            [ 2071] |   printf() {
   2.600 us [ 2071] |   } /* sys_clock_gettime */
   8.200 us [ 2071] |   sys_fstat64();
  20.900 us [ 2071] |   sys_mmap_pgoff();
  20.900 us [ 2071] |   do_page_fault();
            [ 2071] |   sys_clock_gettime() {
   3.800 us [ 2071] |   } /* printf */
  85.800 us [ 2071] | } /* sys_clock_gettime */
   3.700 us [ 2071] | sys_clock_gettime();
```

case 2: VDSO is disabled:
```
root@kesl:~/arm/test# cat /proc/self/maps 
00008000-0000f000 r-xp 00000000 b3:19 160669     /bin/cat
00016000-00017000 r--p 00006000 b3:19 160669     /bin/cat
00017000-00018000 rw-p 00007000 b3:19 160669     /bin/cat
00018000-00039000 rw-p 00000000 00:00 0          [heap]
b6e97000-b6f71000 r-xp 00000000 b3:19 176829     /lib/arm-linux-gnueabihf/libc-2.19.so
b6f71000-b6f78000 ---p 000da000 b3:19 176829     /lib/arm-linux-gnueabihf/libc-2.19.so
b6f78000-b6f7a000 r--p 000d9000 b3:19 176829     /lib/arm-linux-gnueabihf/libc-2.19.so
b6f7a000-b6f7b000 rw-p 000db000 b3:19 176829     /lib/arm-linux-gnueabihf/libc-2.19.so
b6f7b000-b6f7e000 rw-p 00000000 00:00 0 
b6f7e000-b6f95000 r-xp 00000000 b3:19 176819     /lib/arm-linux-gnueabihf/ld-2.19.so
b6f96000-b6f97000 rw-p 00000000 00:00 0 
b6f9b000-b6f9d000 rw-p 00000000 00:00 0 
b6f9d000-b6f9e000 r--p 00017000 b3:19 176819     /lib/arm-linux-gnueabihf/ld-2.19.so
b6f9e000-b6f9f000 rw-p 00018000 b3:19 176819     /lib/arm-linux-gnueabihf/ld-2.19.so
be9e9000-bea0a000 rw-p 00000000 00:00 0          [stack]
bec06000-bec07000 r-xp 00000000 00:00 0          [sigpage]
ffff0000-ffff1000 r-xp 00000000 00:00 0          [vectors]

root@kesl:~/arm/test# uftrace --kernel ./hello
Hello world
# DURATION    TID     FUNCTION
   4.900 us [ 2083] | __cxa_atexit();
   7.400 us [ 2083] | sys_clock_gettime();
   4.900 us [ 2083] | main();
            [ 2083] | sys_clock_gettime() {
            [ 2083] |   printf() {
   4.800 us [ 2083] |   } /* sys_clock_gettime */
  15.300 us [ 2083] |   sys_fstat64();
  39.700 us [ 2083] |   sys_mmap_pgoff();
  19.900 us [ 2083] |   gic_handle_irq();
  41.700 us [ 2083] |   do_page_fault();
            [ 2083] |   sys_clock_gettime() {
   7.400 us [ 2083] |   } /* printf */
 201.600 us [ 2083] | } /* sys_clock_gettime */
   7.400 us [ 2083] | sys_clock_gettime();
```

**Comment by namhyung** [comment]
Thanks for checking this.  Hmm... maybe your system lacks glibc support.  Could you please test below program with `CONFIG_VDSO`?

```
#define _GNU_SOURCE
#include <stdio.h>
#include <string.h>
#include <link.h>

#define VDSO_NAME  "linux-vdso.so.1"

static int check(struct dl_phdr_info *info, size_t sz, void *data)
{
	if (!strcmp(info->dlpi_name, VDSO_NAME))
		printf("VDSO found\n");
	return 0;
}

int main(void)
{
	dl_iterate_phdr(check, NULL);
	return 0;
}
```

**Comment by honggyukim** [comment]
Hi @iamwonseok ,

It seems that the kernel timestamp may not well synced with user timestamp. I saw the similar problem in the other arm board. Can you show me the below command output:
```
# cat /sys/kernel/debug/tracing/trace_clock
```
We expect there is `mono` in the list.

**Comment by honggyukim** [comment]
And can you show me the below output as well:
```
$ cat test.c
#include <unistd.h>
int main()
{
  getuid();
  return 0;
}

$ gcc -pg test.c

$ uftrace -k a.out
```
It must show `sys_getuid()` from kernel output. It's easier function name to check if it's working fine.

**Comment by iamwonseok** [comment]
@namhyung I didn't see any message. I checked the my gblic version: `ldd (Ubuntu EGLIBC 2.19-0ubuntu6) 2.19`.  Hmm... I don't know what the problem is. The reason why I used the ubuntu root file systems on my ARM machine and x86_64 machine. I got normal results on my x86_64 machine. However, a version of both root file systems is the same as ubuntu 14.04. So the glibc version is the same both of ARM and x86_64. 

Anyway, I already ran your program on my x86_64 machine. I didn't see any message, too.

**Comment by iamwonseok** [comment]
Hi @honggyukim here is `trace_clock` log:
```
[local] global counter uptime perf mono mono_raw
```
here is your `test.c` log:
```
# DURATION    TID     FUNCTION
   4.000 us [ 2210] | __cxa_atexit();
   6.000 us [ 2210] | sys_clock_gettime();
   3.900 us [ 2210] | main();
            [ 2210] | sys_clock_gettime() {
            [ 2210] |   getuid() {
   3.900 us [ 2210] |   } /* sys_clock_gettime */
   5.900 us [ 2210] |   sys_getuid();
            [ 2210] |   sys_clock_gettime() {
   5.900 us [ 2210] |   } /* getuid */
  42.600 us [ 2210] | } /* sys_clock_gettime */
   5.900 us [ 2210] | sys_clock_gettime();
```

**Comment by namhyung** [comment]
What's the output of `ldd` then?  The glibc's vdso support can be different for each architecture so some version might support x86 but not arm yet.

Anyway, a simpler solution is always filtering `sys_clock_gettime()` since it won't be called if VDSO support is enabled.  I think we can use `--kernel-full` option to prevent the filtering if users want to.

**Comment by namhyung** [comment]
@honggyukim I think the time is in sync, but the output was messed by the confusing `sys_clock_gettime()`.

**Comment by iamwonseok** [comment]
@namhyung I just checked version of ldd using `ldd --version`.

On ARM machine
```
root@kesl:~/arm# ldd --version
ldd (Ubuntu EGLIBC 2.19-0ubuntu6) 2.19
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

On x86_64 machine
```
wonseok@iamwonseok:~/work$ ldd --version
ldd (Ubuntu EGLIBC 2.19-0ubuntu6.7) 2.19
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

I am building a new root file system by kernel recommendation, because the user who want to use VDSO, the user must have glibc versions 2.2 or higher.

**Comment by namhyung** [comment]
Did you mean glibc version 2.20?  I'd like to see output of `ldd ./hello` then (both of 2.19 and 2.20).

And I think requiring users to upgrade their system does not make sense.  If uftrace can do something, it'd be better for us to do that instead.  What I'm considering is whether applying the filter at record time or replay time..

**Comment by honggyukim** [comment]
> @honggyukim I think the time is in sync, but the output was messed by the confusing sys_clock_gettime().

@namhyung I see. You're right!

**Comment by iamwonseok** [comment]
Yes, if the user wants to use VDSO, kernel recommends the user should have glibc versions 2.2 or higher. I want to construct the environment for getting the similar result comparing to x86_64. So I was building a new root file system which have glibc version  2.2 or higher.

In `make menuconfig ARCH=arm` of kernel:
```
CONFIG_VDSO:                                                                                                        │  
  │                                                                                                                     │  
  │ Place in the process address space an ELF shared object                                                             │  
  │ providing fast implementations of gettimeofday and                                                                  │  
  │ clock_gettime.  Systems that implement the ARM architected                                                          │  
  │ timer will receive maximum benefit.                                                                                 │  
  │                                                                                                                     │  
  │ You must have glibc 2.22 or later for programs to seamlessly                                                        │  
  │ take advantage of this.                                                                                             │  
  │                                                                                                                     │  
  │ Symbol: VDSO [=n]                                                                                                   │  
  │ Type  : boolean                                                                                                     │  
  │ Prompt: Enable VDSO for acceleration of some system calls                                                           │  
  │   Location:                                                                                                         │  
  │     -> System Type                                                                                                  │  
  │   Defined at arch/arm/mm/Kconfig:848                                                                                │  
  │   Depends on: AEABI [=y] && MMU [=y] && CPU_V7 [=y]                                                                 │  
  │   Selects: GENERIC_TIME_VSYSCALL [=n]                                                                               │  
  │                                             
```

@namhyung I am attaching results of `ldd ./hello` and `uftrace --kernel ./hello`

1) using ubuntu zesty root file systems(glibc version 2.24):
```
root@iamwonseok:~/test# ldd ./hello
        linux-vdso.so.1 =>  (0xbee37000)
        libc.so.6 => /lib/arm-linux-gnueabihf/libc.so.6 (0xb6e95000)
        /lib/ld-linux-armhf.so.3 (0x7f632000)
```

```
root@iamwonseok:~/test# uftrace --kernel ./hello
Hello world
# DURATION    TID     FUNCTION
   8.800 us [ 2151] | __cxa_atexit();
            [ 2151] | main() {
            [ 2151] |   printf() {
  15.900 us [ 2151] |     sys_fstat64();
  38.700 us [ 2151] |   } /* printf */
  44.400 us [ 2151] | } /* main */
```

2) using ubuntu trusty root file systems(glibc version 2.19):
```
root@iamwonseok:~/test# ldd ./hello
        linux-vdso.so.1 =>  (0xbea38000)
        libc.so.6 => /lib/arm-linux-gnueabihf/libc.so.6 (0xb6eec000)
        /lib/ld-linux-armhf.so.3 (0x7f599000)
```
```
root@iamwonseok:~/test# uftrace --kernel ./hello
Hello world
# DURATION    TID     FUNCTION
   9.600 us [ 2780] | __cxa_atexit();
  14.600 us [ 2780] | sys_clock_gettime();
   9.500 us [ 2780] | main();
            [ 2780] | sys_clock_gettime() {
            [ 2780] |   printf() {
   9.500 us [ 2780] |   } /* sys_clock_gettime */
  30.000 us [ 2780] |   sys_fstat64();
  80.900 us [ 2780] |   sys_mmap_pgoff();
  40.000 us [ 2780] |   gic_handle_irq();
  82.500 us [ 2780] |   do_page_fault();
            [ 2780] |   sys_clock_gettime() {
  14.700 us [ 2780] |   } /* printf */
 400.700 us [ 2780] | } /* sys_clock_gettime */
  14.700 us [ 2780] | sys_clock_gettime();
```

**Comment by iamwonseok** [comment]
@namhyung I agree with your opinion. By the way, I think that the uftrace needs a detecting the user's environment(CONFIG_VDSO, and glibc version) before the uftrace decides whether the filter is applied or not.

**Comment by namhyung** [comment]
I think we can simply apply filters always on ARM regardless the VDSO support.  If VDSO is usable, the filter will do nothing but it's harmless to have anyway.

**Comment by namhyung** [comment]
@iamwonseok  Any updates?  I think we can simply apply the filter always since x86 won't be affected.  Could you please check `check/filter-vdso` branch?

**Comment by iamwonseok** [comment]
@namhyung Sorry, nowadays I am busy for seeking a job ;D.

Anyway, I checked the `check/filter-vdso` branch and it works fine without `sys_clock_gettime()`. By the way, I think that we want to return the same result whether VDSO is enabled or not. If we want it,  we need to handle additional syscalls such as `sys_rt_sigreturn()`, `sys_gettimeofday()` and `sys_clock_getres()` as the same way of `sys_clock_gettime()` on ARM32.  What do you think about this?

**Comment by namhyung** [comment]
No, the reason why `sys_clock_gettime()` is special is that it's called by uftrace itself (in libmcount.so).  So it doesn't care about ther VDSO functions.

Anyway, good luck with your job seeking!

**Comment by namhyung** [comment]
Merged: 4f12bbf
